src/ai/backend/manager/models/vfolder.py

```diff
@@ -252,179 +265,496 @@ def verify_vfolder_name(folder: str) -> bool:
-                }
-            )
+    queries: list[tuple[Optional[sa.sql.Select], Optional[Callable]]] = []
 
-    entries: List[dict] = []
-    # User vfolders.
     if "user" in allowed_vfolder_types:
-        # Scan my owned vfolders.
-        j = vfolders.join(users, vfolders.c.user == users.c.uuid)
-        query = sa.select(
-            vfolders_selectors + [vfolders.c.permission, users.c.email], use_labels=True
-        ).select_from(j)
-        if not allow_privileged_access or (
-            user_role != UserRole.ADMIN and user_role != UserRole.SUPERADMIN
-        ):
-            query = query.where(vfolders.c.user == user_uuid)
-        await _append_entries(query)
-
-        # Scan vfolders shared with me.
-        j = vfolders.join(
-            vfolder_permissions,
-            vfolders.c.id == vfolder_permissions.c.vfolder,
-            isouter=True,
-        ).join(
-            users,
-            vfolders.c.user == users.c.uuid,
-            isouter=True,
-        )
-        query = (
-            sa.select(
-                vfolders_selectors + [vfolder_permissions.c.permission, users.c.email],
-                use_labels=True,
+        # Query own vfolders
+        queries.append(
+            (
+                _build_query_own_folders(
+                    [vfolders, users.c.email],
+                    user_role,
+                    user_uuid,
+                    extra_vf_conds=extra_vf_conds,
+                ),
+                None,
             )
-            .select_from(j)
-            .where(
-                (vfolder_permissions.c.user == user_uuid)
-                & (vfolders.c.ownership_type == VFolderOwnershipType.USER),
+        )
+        queries.append(
+            (
+                _build_query_user_type_permission_overriden_vfolders(
+                    [vfolders, vfolder_permissions.c.permission, users.c.email],
+                    user_uuid,
+                    extra_vf_conds=extra_vf_conds,
+                ),
+                None,
             )
         )
-        if extra_invited_vf_conds is not None:
-            query = query.where(extra_invited_vf_conds)
-        await _append_entries(query, _is_owner=False)
 
     if "group" in allowed_vfolder_types:
-        # Scan group vfolders.
-        if user_role == UserRole.ADMIN or user_role == "admin":
-            query = (
-                sa.select([groups.c.id])
-                .select_from(groups)
-                .where(groups.c.domain_name == domain_name)
+        # Query project-group vfolders
+        match user_role:
+            case UserRole.SUPERADMIN:
+                group_ids = None
+            case UserRole.ADMIN:
+                group_ids = await query_admin_accessible_groups(db_conn, domain_name)
+            case UserRole.USER:
+                group_ids = await query_user_accessible_groups(db_conn, user_uuid)
+            case _:
+                group_ids = await query_user_accessible_groups(db_conn, user_uuid)
```

Comment:
I think it is safe to remove this branch as `mypy` can check whether all possible branches are exhaustively and explicitly specified or not.